### DISCUSSION TITLE
Store archive of public upstream data in independent repos

### DISCUSSION THREAD

**Discussion by hawkrives** [frog-pond/ccc-server#728] [Ideas]
Today, CCC is roughly structured as

- client requests data from server
- server checks in-memory cache for non-expired data
- if the server finds some, the server responds with the data and asks the client to cache it
- if the server doesn't find any data, it requests it from an upstream service (GitHub pages, BonApp, Google Calendar, a website, etcetera), then stores that response in the server's memory, and passes the body along to the client

I propose that we separate the "ask an upstream" step into a fairly unique infrastructure:

Each upstream should be fetched, parsed, and archived on a schedule. 

CCC's, then, is removed of the need to store an in-memory cache of fetched data to avoid hitting the upstreams too often, and (as a bonus) we get an archive of previous responses. 

---

Each upstream should be represented as two repos: `$upstream-parser` and `$upstream-archive`. 

`$upstream-parser` would be responsible for fetching the data from the upstream and parsing it into ~a structured representation[^1], and pushing it to `$upstream-archive`. It would do this via GitHub Actions on a defined schedule. 

[^1]: ideally using some standard data format, such as RSS/Markdown, vCard, or ICS; my reasoning for this is that we can hopefully avoid needing to change the data representation later, because that affects the ability to look at the history via Git. 

`$upstream-achive`, then, has GitHub pages enabled, and acts as a stand-in for the upstream's server. 

Thoughts?

**Comment by drewvolz** [comment]
I like it!
